cluster-api/apis/cluster/types.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster contains types to represent Kubernetes cluster configuration.
+package cluster // import "k8s.io/kube-deploy/cluster-api/apis/cluster"
```

Comment:
Is it intentional that machine objects are not tied/linked to the cluster object? Could you describe how you are thinking code should get the machine objects of the current cluster?

Reply:
The pattern I believe we settled on for the near-term is that a single cluster should only house a singleton definition for the Cluster object, and only Machine objects that belong to that cluster.

So, a cloud controller should be given the endpoint (and credentials) of an API server to use, and it should just list all Cluster and Machine objects. If zero or multiple Cluster objects are returned, that's an error state, but any number of Machine objects are fair game and all of them should be actuated on.

An additional check you can do for Machines is to attempt to unmarshal the `ProviderConfig` field into the API type you expect to operate on, and ignore the object if unmarshalling fails (or report an error). It's conceivable that even within a single cluster, you may potentially want different controllers handling different kinds of Machines, so the `ProviderConfig` can serve as the differentiator for them.

Reply:
I think the scoping is an important issue that needs to be split out into it's own discussion where that can be the focus. Filed https://github.com/kubernetes/kube-deploy/issues/463

Reply:
Thanks for filing the issue @krousey, will get our team to take a look and we'll add some thoughts in new year.